src/os/bluestore/BlueFS.h

```diff
@@ -265,6 +269,70 @@ class BlueFS {
   struct File : public RefCountedObject {
     MEMPOOL_CLASS_HELPERS();
 
+    /*
+     * WAL v2 files in bluefs have a different format from normal ones. In order to not flush metadata 
+     * for every write we make to data extents, we create a package/envelope around the real data 
+     * that includes length of the data we want to flush and a marker that identifies the flush.
+     *
+     * The format on disk is:
+     * legend: l = length of flush, d = data, m = marker; each character represents one byte
+     *
+     * flush 0 l==24                                     flush 1 l==4             flush 2 l==12
+     * v                                                 v                        v
+     * llll llll dddd dddd dddd dddd dddd dddd mmmm mmmm llll llll dddd mmmm mmmm llll llll dddd dddd dddd mmmm mmmm
+     *
+     */
+    struct wal_flush_t {
+      typedef struct wal_marker_t {
+        uint8_t v[8] = {0};
+      } wal_marker_t;
+      typedef uint64_t wal_length_t;
+      uint64_t file_offset = 0;
+      uint64_t wal_offset = 0; // offset of start of flush, it should be length offset
+      uint32_t header_len = 0;
+      uint32_t tailer_len = 0;
+      uint64_t wal_length = 0;
+
+      static constexpr size_t header_size() {
+        return sizeof(wal_length_t);
+      }
+
+      static constexpr size_t tail_size() {
+        return sizeof(wal_marker_t);
+      }
+
+      uint64_t end_offset() {
+        return get_marker_offset() + tail_size();
+      }
+
+      uint64_t get_payload_offset() {
+        return wal_offset + header_size();
+      }
+
+      uint64_t get_marker_offset() {
+        return get_payload_offset() + wal_length;
+      }
+
+      static constexpr uint64_t extra_envelope_size_on_front_and_tail() {
+        return header_size() + tail_size();
+      }
+
+      static wal_marker_t generate_hashed_marker(uuid_d uuid, uint64_t ino) {
+        wal_marker_t m;
+        uint8_t uuid_copy[16];
+        memcpy(uuid_copy, uuid.bytes(), 16);
+        uint64_t hashed_ino = ino;
+        hashed_ino ^= hashed_ino << 5;
+        hashed_ino ^= hashed_ino << 11;
+        hashed_ino ^= hashed_ino << 23;
+        // use hashed ino in a endiness-agnostic way
+        for (int i = 0; i < 8; i++) {
```

Comment:
8 to be replaced with sizeof(wal_marker) or something

Reply:
I'm not sure with "8", but I added comment to explain the code.

Reply:
AFAIU 8 comes from marker's length. Shouldn't it be sizeof(marker) then?